rust/src/utils/kmeans.rs

```diff
@@ -417,20 +419,39 @@ impl KMeans {
         let dimension = self.dimension;
         let n = data.len() / self.dimension;
         let metric_type = self.metric_type;
-        let cluster_with_distances = stream::iter(0..n)
+        const CHUNK_SIZE: usize = 1024;
+        let cluster_with_distances = stream::iter((0..n).step_by(CHUNK_SIZE))
             // make tiles of input data to split between threads.
-            .chunks(1024)
             .zip(repeat_with(|| (data.clone(), self.centroids.clone())))
-            .map(|(indices, (data, centroids))| async move {
+            .map(|(start_idx, (data, centroids))| async move {
                 let data = tokio::task::spawn_blocking(move || {
-                    let dist = metric_type.batch_func();
+                    let array = data.values();
+                    let centroids_array = centroids.values();
                     let mut results = vec![];
-                    for idx in indices {
-                        let vector = data.slice(idx * dimension, dimension);
-                        let distances = dist(vector.values(), centroids.values(), dimension);
-                        let cluster_id = argmin(distances.as_ref()).unwrap();
-                        let distance = distances.value(cluster_id as usize);
-                        results.push((cluster_id, distance))
+                    for idx in start_idx..min(start_idx + CHUNK_SIZE, n) {
+                        let vector = &array[idx * dimension..(idx + 1) * dimension];
+                        let (cluster_id, distance) = {
+                            let mut min = std::f32::MAX;
+                            let mut min_idx = 0;
+                            for (idx, other) in centroids_array.chunks_exact(dimension).enumerate()
+                            {
+                                // We've found about 40% performance improvement by using static dispatch instead
+                                // of dynamic dispatch.
+                                //
+                                // NOTE: Please make sure run benchmark when changing the following code.
+                                // `RUSTFLAGS="-C target-cpu=native" cargo bench --bench ivf_pq`
+                                let dist = match metric_type {
+                                    MetricType::L2 => vector.l2(other),
+                                    MetricType::Cosine => vector.cosine(other),
+                                };
```

Comment:
I guess the branch prediction will be perfect, but I wonder if we can make this method generic over the distance function then we don't have the branch in the loop at all. (Or maybe this is optimized by the compiler anyways?)

Reply:
Is it possible to do it without passing a function pointer? I tried one with `find_nearest(from: &[f32], to: &[f32], fn: impl Fn(&[f32], &[f32]) -> f32)` which brings it back to 40s (instead of 32s now). 

`vector.l2` and `vector.cosine` are inline function. 

Also i tried hard code `vector.l2(other)` without `match`, it is about 1.5% faster. so i did not push too much for now.

We could prob do a macro to make `match` outside of the loop?

Please advise better way to flatten out the code.

Reply:
I think you should get monomorphization if you change the signature of this function (or make an inner one):

```rust
async fn compute_membership(
    &self,
    data: Arc<Float32Array>,
    distance_fn: impl Fn(&[f32], &[f32]) -> f32
) -> KMeanMembership {
```